In [1]:
import pandas as pd
import numpy as np

### Step 1: Information about the NOC version 1.3 update Dec2019
in this step we will uste the information from the official website and the NOC we will use 2 files for this process. in this steps is mandatory to have internet connection in order to download the oficial documents

In [2]:
noc_structure = pd.read_csv("https://www.statcan.gc.ca/eng/statistical-programs/document/noc-cnp-2016-structure-v4-eng.csv")

In [3]:
noc_structure.shape

(690, 5)

In [4]:
noc_structure.sample(10)

Level Hierarchical structure  Code  \
14       4             Unit group  0114   
678      4             Unit group  9537   
285      4             Unit group  4033   
582      4             Unit group  8411   
684      4             Unit group  9614   
270      2            Major group    34   
637      4             Unit group  9423   
306      4             Unit group  4169   
674      4             Unit group  9533   
31       4             Unit group  0412   

                                           Class title  \
14              Other administrative services managers   
678  Other products assemblers, finishers and inspe...   
285                            Educational counsellors   
582       Underground mine service and support workers   
684       Labourers in wood, pulp and paper processing   
270  Assisting occupations in support of health ser...   
637  Rubber processing machine operators and relate...   
306  Other professional occupations in social scien...   
674      Other wood products assemblers and inspectors   
31   Government managers - economic analysis, polic...   

                                      Class definition  
14   Other administrative services managers plan, o...  
678  Other products assemblers, finishers and inspe...  
285  Educational counsellors advise current and pro...  
582  Underground mine service and support workers p...  
684  Labourers in wood, pulp and paper processing c...  
270  This major group comprises occupations in heal...  
637  Rubber processing machine operators and relate...  
306  Other professional occupations in social scien...  
674  Other wood products assemblers assemble a vari...  
31   Government managers in economic analysis, poli...

In [5]:
noc_elements = pd.read_csv("https://www.statcan.gc.ca/eng/statistical-programs/document/noc-cnp-2016-element-v4-eng.csv")

In [6]:
noc_elements.shape

(43126, 9)

In [7]:
noc_elements.sample(19)

Level  Code                                        Class title  \
13732      4  3232                   Practitioners of natural healing   
41507      4  9537  Other products assemblers, finishers and inspe...   
32300      4  9211          Supervisors, mineral and metal processing   
11185      4  2263  Inspectors in public and environmental health ...   
1654       4   411  Government managers - health and social policy...   
34954      4  9416         Metalworking and forging machine operators   
26828      4  7311  Construction millwrights and industrial mechanics   
35653      4  9421                   Chemical plant machine operators   
11774      4  2282                           User support technicians   
37445      4  9441  Textile fibre and yarn, hide and pelt processi...   
777        4   114             Other administrative services managers   
3562       4   822                           Managers in horticulture   
28188      4  7381                           Printing press operators   
24331      4  7232                                Tool and die makers   
34107      4  9411    Machine operators, mineral and metal processing   
21774      4  6511                Maîtres d'hôtel and hosts/hostesses   
6013       4  1414                                      Receptionists   
38392      4  9461  Process control and machine operators, food an...   
3767       4   912                                 Utilities managers   

      Element Type Label English  \
13732               All examples   
41507               All examples   
32300               All examples   
11185               All examples   
1654                 Main duties   
34954               All examples   
26828               All examples   
35653               All examples   
11774                Main duties   
37445               All examples   
777                 All examples   
3562                Exclusion(s)   
28188               All examples   
24331               All examples   
34107               All examples   
21774                Main duties   
6013                All examples   
38392               All examples   
3767                All examples   

                             Element Description English  Unnamed: 5  \
13732                                         herbalist          NaN   
41507                               pearl earring maker          NaN   
32300        stone products manufacturing foreman/woman          NaN   
11185                quarantine officer - public health          NaN   
1654   This group performs some or all of the followi...         NaN   
34954                 punch press setter - metalworking          NaN   
26828                            oil drill rig mechanic          NaN   
35653        vinyl and asbestos-blending machine tender          NaN   
11774  Participate in the redesign of applications an...         NaN   
37445    winding machine tender - textile manufacturing          NaN   
777        operations manager - administrative services          NaN   
3562   Nursery and greenhouse workers  (See 8432 Nurs...         NaN   
28188               assistant pressman/woman - printing          NaN   
24331                                  bench die sinker          NaN   
34107                                zinc cell operator          NaN   
21774          Order dining room supplies and equipment          NaN   
6013                              hospital receptionist          NaN   
38392                  snuff dryer - tobacco processing          NaN   
3767       transmission operations director - utilities          NaN   

       Unnamed: 6  Unnamed: 7  Unnamed: 8  
13732         NaN         NaN         NaN  
41507         NaN         NaN         NaN  
32300         NaN         NaN         NaN  
11185         NaN         NaN         NaN  
1654          NaN         NaN         NaN  
34954         NaN         NaN         NaN  
26828         NaN         NaN         NaN  
35653         NaN         NaN   

the elements csv file has 9 columns and 43,126 rows and the Stucture file has 5 columns and 690 rows. from the sample exploration, the elements table contains null columns that will be remove before insertion to postgrest


In [8]:
noc_elements = noc_elements.dropna(1)

In [9]:
noc_elements.shape

(43126, 5)

### Step 2: Data Wrangling
In this part we will manage all the data to generated a normalize database in Postgres. 
the main idea is to generate all the necesary Stucture for model Training and Search Engine Support 

In [10]:
noc_structure.columns = ["level","hierarchical_structure","code","title","definition"]

In [11]:
tierOne = noc_structure[noc_structure.level == 1][["code","title","definition"]]
tierOne.columns = ["code_tier_one","title","definition"]

In [12]:
tierTwo = noc_structure[noc_structure.level == 2][["code","title","definition"]]

In [15]:
diff_row_tier = tierTwo[[len(x) > 2 for x in  tierTwo.code]].copy()

In [16]:
data = []
for i in range(len(diff_row_tier)):
    row = diff_row_tier.iloc[i,:].copy()
    start, end = row.code.split("-")
    for n in np.arange(int(start),int(end)+1):
        d = row.copy()
        d.code = str(n).zfill(2)
        data.append(d)

In [17]:
append_data = pd.DataFrame(data)

In [18]:
append_data

code                                              title  \
9    01          Specialized middle management occupations   
9    02          Specialized middle management occupations   
9    03          Specialized middle management occupations   
9    04          Specialized middle management occupations   
9    05          Specialized middle management occupations   
56   07  Middle management occupations in trades, trans...   
56   08  Middle management occupations in trades, trans...   
56   09  Middle management occupations in trades, trans...   

                                           definition  
9   This major group comprises specialized middle ...  
9   This major group comprises specialized middle ...  
9   This major group comprises specialized middle ...  
9   This major group comprises specialized middle ...  
9   This major group comprises specialized middle ...  
56  This major group comprises middle management o...  
56  This major group comprises middle management o...  
56  This major group comprises middle management o...

In [19]:
tierTwo = tierTwo[~tierTwo.code.isin(code_sep.code)].append(append_data)

In [20]:
tierTwo.sort_values("code",inplace=True)

In [21]:
tierTwo.reset_index(inplace= True,drop=True)

In [22]:
tierTwo.columns = ["code_tier_two","title","definition"]
tierTwo.sample(10)

code_tier_two                                              title  \
34            72     Industrial, electrical and construction trades   
14            15  Distribution, tracking and scheduling co-ordin...   
16            22  Technical occupations related to natural and a...   
45            96  Labourers in processing, manufacturing and uti...   
5             05          Specialized middle management occupations   
44            95                        Assemblers in manufacturing   
37            75  Transport and heavy equipment operation and re...   
22            41  Professional occupations in law and social, co...   
19            32                    Technical occupations in health   
41            86  Harvesting, landscaping and natural resources ...   

                                           definition  
34  This major group comprises occupations in indu...  
14  This major group comprises distribution, track...  
16  This major group comprises technical occupatio...  
45  This major group comprises labourers in manufa...  
5   This major group comprises specialized middle ...  
44  This major group comprises assemblers in manuf...  
37  This major group comprises occupations in tran...  
22  This major group comprises professional occupa...  
19  This major group comprises technical occupatio...  
41  This major group comprises labourers in harves...

In [23]:
tierThree = noc_structure[noc_structure.level == 3][["code","title","definition"]]

In [24]:
tierThree.columns = ["code_tier_three","title","definition"]
tierThree.sample(10)

code_tier_three                                              title  \
49              062                Retail and wholesale trade managers   
663             952  Mechanical, electrical and electronics assemblers   
486             728                      Masonry and plastering trades   
74              111  Auditors, accountants and investment professio...   
392             632                                    Chefs and cooks   
491             729                          Other construction trades   
537             745            Longshore workers and material handlers   
259             322        Technical occupations in dental health care   
280             402           College and other vocational instructors   
244             314               Therapy and assessment professionals   

                                            definition  
49   This minor group includes retail and wholesale...  
663  This minor group includes aircraft assemblers ...  
486  This minor group includes bricklayers, concret...  
74   This minor group includes financial auditors, ...  
392  This minor group includes chefs and cooks. The...  
491  This minor group includes roofers, shinglers, ...  
537  This minor group includes longshore workers an...  
259  This minor group includes such technical occup...  
280  This minor group includes instructors who teac...  
244  This minor group includes audiologists, speech...

In [25]:
tierFour = noc_structure[noc_structure.level == 4][["code","title","definition"]]

In [26]:
tierFour.columns = ["code_tier_four","title","definition"]
tierFour.sample(20)

code_tier_four                                              title  \
220           2275  Railway traffic controllers and marine traffic...   
196           2232  Mechanical engineering technologists and techn...   
659           9472  Camera, platemaking and other prepress occupat...   
285           4033                            Educational counsellors   
62            0731                         Managers in transportation   
632           9417                           Machining tool operators   
476           7245          Telecommunications line and cable workers   
672           9531                     Boat assemblers and inspectors   
666           9523  Electronics assemblers, fabricators, inspector...   
43            0511  Library, archive, museum and art gallery managers   
503           7305  Supervisors, motor transport and other ground ...   
269           3237  Other technical occupations in therapy and ass...   
586           8422                  Silviculture and forestry workers   
416           6521                                 Travel counsellors   
481           7252  Steamfitters, pipefitters and sprinkler system...   
477           7246  Telecommunications installation and repair wor...   
644           9436  Lumber graders and other wood processing inspe...   
155           2122                             Forestry professionals   
338           5123                                        Journalists   
364           5242         Interior designers and interior decorators   

                                            definition  
220  Railway traffic controllers co-ordinate passen...  
196  Mechanical engineering technologists and techn...  
659  Camera and platemaking workers operate graphic...  
285  Educational counsellors advise current and pro...  
62   Managers in transportation operations plan, or...  
632  Machining tool operators set up and operate or...  
476  Telecommunications line and cable workers inst...  
672  Boat assemblers assemble wooden, fibreglass an...  
666  Electronics assemblers and fabricators assembl...  
43   Library, archive, museum and art gallery manag...  
503  Supervisors of motor transport and other groun...  
269  Workers in other technical occupations in ther...  
586  Silviculture and forestry workers perform a va...  
416  Travel counsellors advise clients on travel op...  
481  Steamfitters and pipefitters lay out, assemble...  
477  Telecommunications installation and repair wor...  
644  Lumber graders and other wood processing inspe...  
155  Forestry professionals conduct research, devel...  
338  Journalists research, investigate, interpret a...  
364  Interior designers and interior decorators con...

In [44]:
tierFour

code_tier_four                                              title  \
3             0011                                        Legislators   
4             0012           Senior government managers and officials   
5             0013  Senior managers - financial, communications an...   
6             0014  Senior managers - health, education, social an...   
7             0015  Senior managers - trade, broadcasting and othe...   
..             ...                                                ...   
685           9615  Labourers in rubber and plastic products manuf...   
686           9616                    Labourers in textile processing   
687           9617          Labourers in food and beverage processing   
688           9618           Labourers in fish and seafood processing   
689           9619  Other labourers in processing, manufacturing a...   

                                            definition  
3    Legislators participate in the activities of a...  
4    Senior government managers and officials plan,...  
5    Senior managers in financial, communications a...  
6    Senior managers in health, education, social a...  
7    Senior managers in trade, broadcasting and oth...  
..                                                 ...  
685  Labourers in rubber and plastic products manuf...  
686  Labourers in textile processing perform a vari...  
687  Labourers in food and beverage processing perf...  
688  Labourers in fish and seafood processing perfo...  
689  Other labourers in processing, manufacturing a...  

[500 rows x 3 columns]

In [28]:
noc_elements.columns = ["level","code","title","type","description"]

In [29]:
noc_elements["code_tier_four"] = noc_elements.code.apply(lambda t: str(t).zfill(4))
noc_elements["code_tier_three"] = noc_elements.code_tier_four.apply(lambda t4: t4[:3] )
noc_elements["code_tier_two"] = noc_elements.code_tier_four.apply(lambda t4: t4[:2] )
noc_elements["code_tier_one"] = noc_elements.code_tier_four.apply(lambda t4: t4[:1] )

In [31]:
noc_elements[noc_elements.code == 2121]

level  code                              title                     type  \
7685      4  2121  Biologists and related scientists  Illustrative example(s)   
7686      4  2121  Biologists and related scientists  Illustrative example(s)   
7687      4  2121  Biologists and related scientists  Illustrative example(s)   
7688      4  2121  Biologists and related scientists  Illustrative example(s)   
7689      4  2121  Biologists and related scientists  Illustrative example(s)   
...     ...   ...                                ...                      ...   
7854      4  2121  Biologists and related scientists              Main duties   
7855      4  2121  Biologists and related scientists  Employment requirements   
7856      4  2121  Biologists and related scientists  Employment requirements   
7857      4  2121  Biologists and related scientists  Employment requirements   
7858      4  2121  Biologists and related scientists  Employment requirements   

                                            description code_tier_four  \
7685                                         anatomist            2121   
7686                                    bacteriologist            2121   
7687                                  bioinformatician            2121   
7688                                         biologist            2121   
7689                                          botanist            2121   
...                                                 ...            ...   
7854  Biologists and related scientists may speciali...           2121   
7855  A bachelor's degree in biology or in a related...           2121   
7856  A master's or doctoral degree in biology or a ...           2121   
7857  Post-doctoral research experience is usually r...           2121   
7858  Biologists are regulated in Alberta and Britis...           2121   

     code_tier_three code_tier_two code_tier_one  
7685             212            21             2  
7686             212            21             2  
7687             212            21             2  
7688             212            21             2  
7689             212            21             2  
...              ...           ...           ...  
7854             212            21             2  
7855             212            21             2  
7856             212            21             2  
7857             212            21             2  
7858             212            21             2  

[174 rows x 9 columns]

In [35]:
information_type = noc_elements["type"].drop_duplicates().to_list()

In [40]:
#for i_type in information_type:
#    noc_elements[noc_elements.type == i_type][["code_tier_four","description"]].sample(5)

In [58]:
type_to_table = {t:"" for t in noc_elements.type.drop_duplicates().to_list()}

In [62]:
type_to_table["Illustrative example(s)"] = "NOC_EXAMPLES"
type_to_table["All examples"] = "NOC_EXAMPLES"
type_to_table["Exclusion(s)"] = "NOC_EXCLUSIONS"
type_to_table["Main duties"] = "NOC_MAIN_DUTIES"
type_to_table["Employment requirements"] = "NOC_EMPLOYMENT_REQUIREMENTS"
type_to_table["Inclusion(s)"] = "NOC_INCLUSIONS"
type_to_table["Additional information"] = "NOC_ADITIONAL_INFORMATION"




In [64]:
type_to_table

{'Illustrative example(s)': 'NOC_EXAMPLES',
 'All examples': 'NOC_EXAMPLES',
 'Exclusion(s)': 'NOC_EXCLUSIONS',
 'Main duties': 'NOC_MAIN_DUTIES',
 'Employment requirements': 'NOC_EMPLOYMENT_REQUIREMENTS',
 'Inclusion(s)': 'NOC_INCLUSIONS',
 'Additional information': 'NOC_ADITIONAL_INFORMATION'}

In [66]:
noc_elements[noc_elements.type == "Main duties"][["code_tier_four","description"]]

code_tier_four                                        description
52              0011  This group performs some or all of the followi...
53              0011       Enact, amend or repeal laws and regulations 
54              0011  Participate in developing or amending governme...
55              0011  Represent their government at local, national ...
56              0011  Respond to matters of concern to constituents ...
...              ...                                                ...
43118           9619            Check and weigh materials and products 
43119           9619  Sort, pack, crate and package materials and pr...
43120           9619  Assist machine operators, assemblers and other...
43121           9619                    Clean work areas and equipment 
43122           9619  Perform other labouring and elemental activiti...

[5867 rows x 2 columns]

In [67]:
noc_elements[noc_elements.type == "Additional information"][["code_tier_four","description"]]

code_tier_four                                        description
159             0012  There is mobility among senior management occu...
248             0013  There is mobility among senior management occu...
358             0014  There is mobility among senior management occu...
435             0015  There is mobility among senior management occu...
511             0016  There is mobility among senior management occu...
...              ...                                                ...
42687           9617  There is some mobility among occupations in th...
42688           9617  Labourers in food and beverage processing may ...
42735           9618  Progression to other occupations in the fish p...
43124           9619  There is some mobility among occupations in th...
43125           9619  Progression to machine operating positions is ...

[629 rows x 2 columns]